### ISSUE TITLE
FeatReq: Allow patching dependencies

### ISSUE THREAD

**Issue by voidus** [python-poetry/poetry#7717]
Hi everyone!

- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [FAQ](https://python-poetry.org/docs/faq/) and general [documentation](https://python-poetry.org/docs/) and believe that my question is not already covered.

## Feature Request
Sometimes, I encounter a small issue in one of my dependencies. Being a good citizen, I fix it and open a PR. Given that maintainers need a while, that doesn't solve my immediate problem though.

In the past, I've dealt with it like this:
- Point the dependency at my fork
- Try really hard to remember to point it back to upstream once my PR is merged

This has a few issues
- My brain is like a chipmunk and I forget
- Upstream might nitpick or flat-out reject my PR (as is their right)

Instead, it would be great if I could add a `.patch` file to my repository, which is then applied to the dependency. Conveniences around this could include:

- Automatic check if my PR is merged
- Warnings/errors if the dependency is updated while my patch exists, even though it still applies

But I think a simple version could look like this:

```
[tool.poetry.group.dev.dependencies]
pytest-testmon = { version = "^2.0.0", patches = ["./patches/pytest-testmon-pr-212.patch"] }
```

I see poetry applying that patch on install and erroring out if it doesn't apply.

We might also want to adjust commands that print versions to add like a * or `[p]` after the version if it was patched, idk.

**Comment by dimbleby** [comment]
there are rather a lot of well-meaning feature requests in this repository which, realistically, are unlikely ever to be implemented.  See the long list of open issues.

Meanwhile, there are plenty of more standard mechanisms for providing your own patch: point at a git repository, point at a local directory, publish your own variant of the package (eg with a new name, or using a local version).

(And I don't see why your chipmunk brain is likely to do any better with your proposal than with one of the existing approaches!)

Suggest using one of those, and not holding your breath for any change here.

**Comment by voidus** [comment]
Thank you for the quick response.

To be clear, I wasn't expecting this to be implemented anytime soon (or at all, really, even though I think it solves a real problem!)

All your proposals include pointing my dependency at another source, which then all make me miss further upgrades and require me to pay particular attention on this specific dependency.

My proposal would be different in that it allows users to keep their dependency pointed at upstream, thus not missing updates. Once the patch becomes unnecessary, it will stop applying cleanly and the error will grab attention, thus relieving the user of having to stay aware of the issue.

PS: I know you didn't mean it like that, but me joking about myself doesn't usually make it appropriate for other people to do it as well, at least as long as we don't know each other. As I said, I know that was not your intent, just wanted to point that out.

**Comment by dimbleby** [comment]
No offence meant, we all are chipmunks.

IMO in 99.99% of possible worlds, this isn't going to happen.  It's just a question of whether we admit that now, or later.

I could be wrong - I've been wrong before! - so if you want to leave this open, then good luck.

**Comment by Secrus** [comment]
I am going to close it as far beyond what Poetry should include. In such cases, you can look towards tools like `vendoring` and vendor the dependency you need patched.

**Comment by kshpytsya** [comment]
Would it be possible to implement something like this as a Poetry plugin?
